dana/lerobot-community-datasets (#2839)
huggingface/blog: lerobot-datasets.md

---
title: "LeRobot Community Datasets: The “ImageNet” of Robotics — When and How?" 
thumbnail: /blog/assets/195_lerobot_datasets/1.png
authors:
- user: danaaubakirova
- user: Beegbrain
  guest: true
- user: mshukor
- user: m1b
  guest: true
- user: villekuosmanen
  guest: true
- user: cadene
---

> Recent advances in Vision-Language-Action (VLA) models have enabled robots to perform a wide range of tasks—from simple commands like “grasp the cube” to more complex activities like folding laundry or cleaning a table. These models aim to achieve **generalization**: the ability to perform tasks in novel settings, with unseen objects, and in varying conditions.

Reviewer: I'd maybe suggest to include a short tldr in the very beginning to explain what this is about: we are trying to empower the community to build better robotics datasets and after a few months we have great contributions and have identified a few guidelines to keep progressing.

Author: done! good idea!

## Why Robotics Lacks Its ImageNet Moment?

So far, the majority of robotics datasets come from structured academic environments. Even if we scale up to millions of demonstrations, one dataset will often dominate, limiting diversity. Unlike ImageNet—which aggregated internet-scale data and captured the real world more holistically—robotics lacks a comparably diverse, community-driven benchmark.

This is largely because collecting data for robotics requires **physical hardware and significant effort**.

## Building a LeRobot Community

That’s why, at **LeRobot**, we’re working to make robotics data collection more accessible—at home, at school, or anywhere. We're:

- Simplifying the recording pipeline  
- Streamlining uploading to the Hugging Face Hub  
- Reducing hardware costs  

We're already seeing the results: the number of community-contributed datasets on the Hub is growing rapidly.
<div align="center">

>   <img src="https://cdn-uploads.huggingface.co/production/uploads/640e21ef3c82bd463ee5a76d/9E7qfkq1sxMJcxecSrJDN.webp" alt="Growth of lerobot datasets" width="500"/>

Reviewer: (figure nit: maybe title it as "LeRobot datasets uploaded over time", or use the same caption as below)

Author: done!

>   <p><b>Figure 1:</b> Growth of <i>lerobot</i> datasets on the Hugging Face Hub over time.</p>

Reviewer: ```suggestion
  <p>Growth of <i>lerobot</i> datasets on the Hugging Face Hub over time.</p>
```

I don't think we need named references for the figures, it may read too formal :)

</div>

> Figure 1 shows the steady growth of the *lerobot* datasets on the Hub.  
> Figure 2 breaks them down by robot type—highlighting contributions from So100 and Koch as dominant platforms. This momentum brings us closer to a future where datasets reflect a **global effort**, not just one lab or institution.

Reviewer: ```suggestion
Not only is the growth of datasets significant over the past few months, but if we break them down by robot type we also notice that most contributions are to So100 and Koch as the dominant platforms. This momentum brings us closer to a future where datasets reflect a **global effort**, not just one lab or institution.
```

Why are So100 and Koch important?

Reviewer: Can we mention a couple of notable datasets that have been contributed? I think it could be a cool way to nod the community.

<div align="center">
  <img src="https://cdn-uploads.huggingface.co/production/uploads/640e21ef3c82bd463ee5a76d/rhfdwybDuIu2ULGs7Nb7E.webp" alt="Distribution of lerobot datasets" width="500"/>

>   <p><b>Figure 2:</b> Distribution of <i>lerobot</i> datasets by robot type.</p>

Reviewer: ```suggestion
  <p>Distribution of <i>lerobot</i> datasets by robot type.</p>
```

</div>

## Scaling Responsibly

As robotics data collection becomes more democratized, **curation becomes the next challenge**. While these datasets are still collected in constrained setups, they are a crucial step toward affordable, general-purpose robotic policies. Not everyone has access to expensive hardware—but with **shared infrastructure and open collaboration**, we can build something far greater.

> 🧠 **“Generalization isn’t solved in a lab—it’s taught by the world.”**  
> The more diverse our data, the more capable our models will be.

> ## Why This Blogpost?

Reviewer: I'd maybe remove this section or move it closer to the beginning of the post.

Author: yes! done

The goal of this blogpost is to:

- Recognize the growing impact of community-contributed LeRobot datasets  
- Identify current challenges  
- Outline practical steps to maximize the value of this collective effort  

---

> # Better data = Better models

Reviewer: I think it may look a bit weird to have big title headings in the middle of the post.

Author: removed

Within this framework, efficient real-world data collection is indispensable—it anchors learned behaviors in actual robotic hardware and **closes the sim-to-real gap**, ultimately improving the generalization, adaptability, and performance of robotics foundation models.

By expanding the **volume and diversity of real-world datasets**, we reduce fragmentation between heterogeneous data sources. When datasets are disjoint in terms of environment, embodiment, or task distribution, models struggle to transfer knowledge across domains.

> 🔗 **Real-world data acts as connective tissue**—it aligns abstract priors with grounded action and enables the model to build more coherent and transferable representations.

As a result, increasing the proportion of real robot interactions does not merely enhance realism—it **structurally reinforces** the links between all layers of the pyramid, leading to more robust and capable policies.

<div align="center">
  <img src="https://cdn-uploads.huggingface.co/production/uploads/640e21ef3c82bd463ee5a76d/eBmRnO1MsJ5SLxo1pMStf.png" alt="Data Pyramid for Robot Foundation Model Training" width="500"/>

>   <p><b>Figure 3:</b> Data Pyramid for Robot Foundation Model Training. Adapted from <a href="https://arxiv.org/pdf/2503.14734">Gr00t</a> (Yang et al., 2025). Data quantity decreases while embodiment specificity increases from bottom to top.</p>

Reviewer: ```suggestion
  <p>Data Pyramid for Robot Foundation Model Training. Adapted from <a href="https://arxiv.org/pdf/2503.14734">Gr00t</a> (Yang et al., 2025). Data quantity decreases while embodiment specificity increases from bottom to top.</p>
```

</div>

---
# Challenges with Current Community Datasets

At LeRobot, we’ve started developing an automatic curation pipeline to post-process community datasets. During the post-processing phase, we’ve identified several areas where improvements can further boost dataset quality and facilitate more effective curation going forward:

### 1. Incomplete or Inconsistent Task Annotations

Many datasets lack task descriptions, lack details or are ambiguous in the task to be done. Semantics is currently at the core of cognition, meaning that understanding the context and specifics of a task is crucial for robotic performance. Detailed expressions ensure that robots understand exactly what is expected, but also provide a broader knowledge and vocabulary to the cognition system. Ambiguity can lead to incorrect interpretation and, consequently, incorrect actions.

Task instructions can be:
- Empty
- Too short (e.g. “Hold”, “Up”)
- Without any specific meaning (e.g. “task desc”, “desc”)

Subtask-level annotations are often missing, making it difficult to model complex task hierarchies.  

> While this can be handled with VLM, it is still better to have a task annotation provided by the author of the dataset at hand.

Reviewer: ```suggestion
While this can be handled with VLMs, it is still better to have a task annotation provided by the author of the dataset at hand.
```

Author: accepted

Reviewer 2: I think this is still not applied 😄

### 2. Feature Mapping Inconsistencies

Features like `images.laptop` are ambiguously labeled:
- Sometimes it's a third-person view
- Other times it's more like a gripper (wrist) camera

Manual mapping of dataset features to standardized names is time-consuming and error-prone.  
We can possibly automate feature type inference using VLMs or computer vision models to classify camera perspectives. However, keeping this in mind helps to have a cleaner dataset.

### 3. Low-Quality or Incomplete Episodes

Some datasets contain:
- Episodes with only 1 or very few frames
- Manually deleted data files (e.g., deleted `.parquet` files without reindexing), breaking the sequential consistency.

### 4. Inconsistent Action/State Dimensions

Different datasets use different action or state dimensions, even for the same robot (e.g., `so100`).  
Some datasets show inconsistencies in action/state format.

---

# What Makes a Good Dataset?

> Now that we know that creating a high-quality dataset is essential for training reliable and generalizable robot policies. To assist you in collecting effective data, we’ve outlined a checklist of best practices across four key areas:

Reviewer: ```suggestion
Now that we know that creating a high-quality dataset is essential for training reliable and generalizable robot policies, we have outlined a checklist of best practices to assist you in collecting effective data.
```

Author: accepted

## Image Quality

> - ✅ Use **at least two camera views**

Reviewer 3: ```suggestion
- ✅ Use preferably **two camera views**
```

- ✅ Ensure **steady video capture** (no shaking)
- ✅ Maintain **neutral, stable lighting** (avoid overly yellow or blue tones)
- ✅ Ensure **consistent exposure** and **sharp focus**
- ✅ **Leader arm should not appear** in the frame
- ✅ The **only moving objects** should be the follower arm and the manipulated items (avoid human limbs/bodies)
- ✅ Use a **static, non-distracting background**, or apply controlled variations

> - ✅ Record in **high resolution** (at least 720p)

Reviewer 3: Maybe?
```suggestion
- ✅ Record in **high resolution** (at least 480x640 / 720p)
```

Author: yes, I accepted the changes! thanks!

## Metadata & Recording Protocol

> - ✅ Select the **correct robot type** in the metadata

Reviewer 3: What does it mean in practice?

Author: I think Marina meant during recording you need to pass correct robot type in the arguments.

- ✅ Record videos at approximately **30 frames per second (FPS)**

> - ✅ If **deleting episodes**, make sure to **update the metadata files accordingly**

Reviewer 3: ```suggestion
- ✅ If **deleting episodes**, make sure to **update the metadata files accordingly** (we will provide proper tools to edit datasets)
```

## Feature Naming Conventions

Use a consistent and interpretable naming scheme for all camera views and observations:

**Format:**
```bash
<modality>.<location>
```

**Examples:**

- `images.top`
- `images.front`
- `images.left`
- `images.right`

**Avoid device-specific names:**

- ❌ `images.laptop`

> - ❌ `images.phone`

Reviewer 3: That's what we advise in our tutorial :')
Not sure about this. We should be invariant to camera position.

Author: This isn't so much about camera position, but rather about standardising feature names across datasets — for example, images.phone corresponds more to a third-person view or a gripper camera? We're currently doing this mapping manually, but as we scale to hundreds of datasets, that will become impractical.
cc: @mshukor

**For wrist-mounted cameras, specify orientation:**

- `images.wrist.left`
- `images.wrist.right`
- `images.wrist.top`

> - `images.wrist.bottom`

Reviewer 3: same, I am not sure about this -- it's super niche

Author: Yes, I agree. But this is a helpful addition. it gives a clear naming direction instead of leaving people unsure how to label these kinds of observations. wdyt?

Reviewer: I think it's useful information when you are in the process of building a dataset, but maybe too much detail for an introductory post. People that get excited and want to contribute will use the rest of the resources. Perhaps we can write it in a more general way without providing all the labels and orientations.

> Consistent naming improves clarity and helps downstream models better interpret spatial configurations and multi-view inputs.

## Task Annotation

> - ✅ Use the `task` field to **clearly describe the robot’s objective**

Reviewer: ```suggestion
- ✅Try to **clearly describe the robot’s objective**
```

(in the same vein as above, the field name doesn't matter too much)

Author: Thanks for the suggestion! I'd prefer to keep it as is, since it's closely tied to the recording pipeline and specs. Specifying clearly the field is crucial — many issues with the current dataset stem from them being unclear or overlooked.

  - *Example:* `Pick the yellow lego block and put it in the box`
- ✅ Keep task descriptions **concise** (between **25–50 characters**)
- ✅ Avoid vague or generic names like `task1`, `demo2`, etc.

---

Below, we provide a checklist that serves as a guideline for recording datasets, outlining key points to keep in mind during the data collection process.

<div align="center">
  <img src="https://cdn-uploads.huggingface.co/production/uploads/640e21ef3c82bd463ee5a76d/6bZ8VEU-kgpDjLCokHj5J.png" alt="Dataset Recording Checklist" width="600"/>
  <p><b>Figure 4:</b> Dataset Recording Checklist – a step-by-step guide to ensure consistent and high-quality real-world data collection.</p>
</div>

> > Start recording, start contributing—because the future of generalist robots depends on the data we build today.

Reviewer: I'd put this in its own section (for instance, "How can you help?"), and try to make it a bit more exciting :)

Author: Done!

> > Start recording, start contributing—because the future of generalist robots depends on the data we build today.

Reviewer: I'd put this in its own section (for instance, "How can you help?"), and try to make it a bit more exciting :)

## Additional Resources

- [LeRobot](https://huggingface.co/lerobot)
- [Pi0 and Pi0-FAST in PyTorch](https://huggingface.co/blog/pi0)
- [Pi0 Blogpost](https://www.physicalintelligence.company/blog/pi0)

> - [FAST: Efficient Robot Action Tokenization](https://www.physicalintelligence.company/research/fast)

Reviewer: Not sure if all of these are relevant. Perhaps we can refer people to hf.co/lerobot as part of the call-to-action section above and that should be enough to get started.

## Scaling Responsibly

As robotics data collection becomes more democratized, **curation becomes the next challenge**. While these datasets are still collected in constrained setups, they are a crucial step toward affordable, general-purpose robotic policies. Not everyone has access to expensive hardware—but with **shared infrastructure and open collaboration**, we can build something far greater.

> 🧠 **“Generalization isn’t solved in a lab—it’s taught by the world.”**  
> The more diverse our data, the more capable our models will be.

---
## Better data = Better models

Why does data quality matter? Poor-quality data results in poor downstream performance, biased outputs, and models that fail to generalize. Hence, **efficient and high-quality data collection** plays a critical role in advancing generalist robotic policies.

While foundation models in vision and language have thrived on massive, web-scale datasets, robotics lacks an “Internet of robots”—a vast, diverse corpus of real-world interactions. Instead, robotic data is fragmented across different embodiments, sensor setups, and control modes, forming isolated *data islands*.

> To overcome this, recent approaches like [Gr00t](https://arxiv.org/pdf/2503.14734) organize training data as a **pyramid**, where:

Reviewer 2: ```suggestion
To overcome this, recent approaches like [Gr00t](https://huggingface.co/papers/2503.14734) organize training data as a **pyramid**, where:
```

Within this framework, efficient real-world data collection is indispensable—it anchors learned behaviors in actual robotic hardware and **closes the sim-to-real gap**, ultimately improving the generalization, adaptability, and performance of robotics foundation models.

By expanding the **volume and diversity of real-world datasets**, we reduce fragmentation between heterogeneous data sources. When datasets are disjoint in terms of environment, embodiment, or task distribution, models struggle to transfer knowledge across domains.

> 🔗 **Real-world data acts as connective tissue**—it aligns abstract priors with grounded action and enables the model to build more coherent and transferable representations.

As a result, increasing the proportion of real robot interactions does not merely enhance realism—it **structurally reinforces** the links between all layers of the pyramid, leading to more robust and capable policies.

<div align="center">
  <img src="https://cdn-uploads.huggingface.co/production/uploads/640e21ef3c82bd463ee5a76d/eBmRnO1MsJ5SLxo1pMStf.png" alt="Data Pyramid for Robot Foundation Model Training" width="500"/>

>   <p>Data Pyramid for Robot Foundation Model Training. Adapted from <a href="https://arxiv.org/pdf/2503.14734">Gr00t</a> (Yang et al., 2025). Data quantity decreases while embodiment specificity increases from bottom to top.</p>

Reviewer 2: ```suggestion
  <p>Data Pyramid for Robot Foundation Model Training. Adapted from <a href="https://huggingface.co/papers/2503.14734">Gr00t</a> (Yang et al., 2025). Data quantity decreases while embodiment specificity increases from bottom to top.</p>
```
